spec/src/main/asciidoc/query-language.asciidoc

```diff
@@ -0,0 +1,341 @@
+- paths,
+- function calls, and
+- operator expressions.
+
+==== Literal expressions
+
+A string, integer, or decimal literal is assigned the type it would be assigned in Java. So, for example, `'Hello'` is assigned the type `java.lang.String`, `123` is assigned the type `int`, `1e4` is assigned the type `double`, and `1.23f` is assigned the type `float`.
+
+The syntax for literal expressions is given by the `literal` grammar rule, and in the previous section titled <<Lexical structure>>.
+
+When executed, a literal expression evaluates to its literal value.
+
+==== Special values
+
+The special values `true` and `false` are assigned the type `boolean`, and evaluate to their literal values.
+
+The special values `local date` and `local datetime` are assigned the types `java.time.LocalDate` and `java.time.LocalDateTime`, and evaluate to the current date and current datetime of the database server, respectively.
+
+The syntax for special values is given by the `special_expression` grammar rule.
+
+==== Parameter expressions
+
+A parameter expression, with syntax given by `input_parameter`, is assigned the type of the repository method parameter it matches. For example, the parameter `:titlePattern` is assigned the type `java.lang.String`:
+
+[source,java]
+----
+@Query("where title like :titlePattern")
+List<Book> booksMatchingTitle(String titlePattern);
+----
+
+When executed, a parameter expression evaluates to the argument supplied to the parameter of the repository method.
+
+==== Enum literals
+
+An _enum literal expression_ is a Java identifier, with syntax specified by `enum_literal`, and may only occur as the right operand of a `set` assignment or `=`/`<>` equality comparison. It is assigned the type of the left operand of the assignment or comparison. The type must be a Java `enum` type, and the identifier must be the name of an enumerated value of the `enum` type. For example, `day <> MONDAY` is a legal comparison expression.
+
+When executed, an enum expression evaluates to the named member of the Java `enum` type.
+
+==== Path expressions
+
+A _path expression_ is a period-separated list of Java identifiers, with syntax specified by `state_field_path_expression`. Each identifier is interpreted as the name of a field of an entity or embeddable class. Each prefix of the list is assigned a Java type:
+
+- the first element of the list is assigned the type of the named field of the entity being queried, and
+- each subsequent element is assigned the type of the named field of the type assigned to the previous element.
+
+The type of the whole path expression is the type of the last element of the list. For example, `pages` is assigned the type `int`, `address` is assigned the type `org.example.Address`, and `address.street` is assigned the type `java.lang.String`.
+
+When executed, a path expression is evaluated in the context of a given record of the queried entity type, and evaluates to the value of the entity field for the given record.
```

Comment:
what happens when something in the path is `null`? Do we support `null` safe dereferencing?

Reply:
This is something I need to clarify.

Reply:
@graemerocher see 34ca9693c2bb8fce5c2b3e9c84dd69644de008f0.